rust/template/d3log/src/forwarder.rs

```diff
@@ -0,0 +1,137 @@
+                }
+            }
+        }
+    }
+}
+
+#[derive(Clone)]
+struct Entry {
+    port: Option<Port>,
+    batches: VecDeque<Batch>,
+    registrations: VecDeque<Node>,
+}
+
+pub struct Forwarder {
+    eval: Evaluator,
+    fib: Arc<Mutex<HashMap<Node, Arc<Mutex<Entry>>>>>,
+}
+
+impl Forwarder {
+    pub fn new(eval: Evaluator, dispatch: Arc<Dispatch>, _management: Port) -> Arc<Forwarder> {
+        let f = Arc::new(Forwarder {
+            eval: eval.clone(),
+            fib: Arc::new(Mutex::new(HashMap::new())),
+        });
+        dispatch
+            .clone()
+            .register(
+                "d3_application::Forward",
+                Arc::new(ForwardingEntryHandler {
+                    eval: eval.clone(),
+                    forwarder: f.clone(),
+                }),
+            )
+            .expect("register");
+        f
+    }
+
+    fn lookup(&self, n: Node) -> Arc<Mutex<Entry>> {
+        self.fib
+            .lock()
+            .expect("lock")
+            .entry(n)
+            .or_insert_with(|| {
+                Arc::new(Mutex::new(Entry {
+                    port: None,
+                    batches: VecDeque::new(),
+                    registrations: VecDeque::new(),
+                }))
+            })
+            .clone()
+    }
+
+    pub fn register(&self, n: Node, p: Port) {
+        // overwrite warning?
+        let entry = self.lookup(n);
+        {
+            entry.lock().expect("lock").port = Some(p.clone());
+        }
+
+        while let Some(b) = { entry.lock().expect("lock").batches.pop_front() } {
+            p.clone().send(b);
+        }
+        while let Some(r) = { entry.lock().expect("lock").registrations.pop_front() } {
+            self.register(r, p.clone());
+        }
+    }
+}
+
+impl Transport for Forwarder {
+    fn send(&self, b: Batch) {
+        let mut output = HashMap::<Node, ValueSet>::new();
+        for (rel, v, weight) in &ValueSet::from(self.eval.clone(), b.clone()).expect("iterator") {
+            if let Some((loc_id, in_rel, inner_val)) = self.eval.localize(rel, v.clone()) {
+                output
+                    .entry(loc_id)
+                    .or_insert_with(|| ValueSet::new(self.eval.clone()))
+                    .insert(in_rel, inner_val, weight);
+            }
+        }
+
+        for (nid, localized_batch) in output.drain() {
+            let p = {
+                match self.lookup(nid).lock() {
+                    Ok(mut x) => match &x.port {
+                        Some(x) => x.clone(),
+                        None => {
+                            x.batches.push_front(Batch {
+                                metadata: b.metadata.clone(),
+                                body: BatchBody::Value(localized_batch),
+                            });
+                            break;
+                        }
+                    },
+                    Err(_) => panic!("lock"),
+                }
+            };
```

Comment:
```suggestion
            let port = {
                let entry = self.lookup(nid).lock().expect("poisoned lock");

                if let Some(port) = entry.port {
                    port
                } else {
                    entry.batches.push_front(Batch {
                        metadata: b.metadata.clone(),
                        body: BatchBody::Value(localized_batch),
                    });

                    break;
                }
            };
```
Using the `if let` should let you elide a clone